server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java

```diff
@@ -745,4 +762,196 @@ private static GetSnapshotsRequestBuilder baseGetSnapshotsRequest(String[] repoN
+        final var expectedOffset = Math.min(selectedSnapshots.size(), skippedByFromSortValue + getSnapshotsRequest.offset());
+        final var expectedSize = Math.min(
+            selectedSnapshots.size() - expectedOffset,
+            getSnapshotsRequest.size() == GetSnapshotsRequest.NO_LIMIT ? Integer.MAX_VALUE : getSnapshotsRequest.size()
+        );
+
+        // get the actual response
+        final GetSnapshotsResponse getSnapshotsResponse = safeAwait(
+            l -> client().execute(TransportGetSnapshotsAction.TYPE, getSnapshotsRequest, l)
+        );
+
+        // verify it returns the expected results
+        assertEquals(
+            selectedSnapshots.stream().skip(expectedOffset).limit(expectedSize).map(SnapshotInfo::snapshotId).toList(),
+            getSnapshotsResponse.getSnapshots().stream().map(SnapshotInfo::snapshotId).toList()
+        );
+        assertEquals(expectedSize, getSnapshotsResponse.getSnapshots().size());
+        assertEquals(selectedSnapshots.size() - skippedByFromSortValue, getSnapshotsResponse.totalCount());
+        assertEquals(selectedSnapshots.size() - expectedOffset - expectedSize, getSnapshotsResponse.remaining());
+        assertEquals(getSnapshotsResponse.remaining() > 0, getSnapshotsResponse.next() != null);
+
+        // now use ?after to page through the rest of the results
+        var nextRequestAfter = getSnapshotsResponse.next();
+        var nextExpectedOffset = expectedOffset + expectedSize;
+        var remaining = getSnapshotsResponse.remaining();
+        while (nextRequestAfter != null) {
+            final var nextSize = between(1, remaining);
+            final var nextRequest = new GetSnapshotsRequest(TEST_REQUEST_TIMEOUT, requestedRepositories, requestedSnapshots)
+                // same name filters, same ?sort and ?order params, new ?size, but no ?offset or ?from_sort_value because of ?after
+                .sort(sortKey)
+                .order(order)
+                .size(nextSize)
+                .after(SnapshotSortKey.decodeAfterQueryParam(nextRequestAfter));
+            final GetSnapshotsResponse nextResponse = safeAwait(l -> client().execute(TransportGetSnapshotsAction.TYPE, nextRequest, l));
+
+            assertEquals(
+                selectedSnapshots.stream().skip(nextExpectedOffset).limit(nextSize).map(SnapshotInfo::snapshotId).toList(),
+                nextResponse.getSnapshots().stream().map(SnapshotInfo::snapshotId).toList()
+            );
+            assertEquals(nextSize, nextResponse.getSnapshots().size());
+            assertEquals(selectedSnapshots.size(), nextResponse.totalCount());
+            assertEquals(remaining - nextSize, nextResponse.remaining());
+            assertEquals(nextResponse.remaining() > 0, nextResponse.next() != null);
+
+            nextRequestAfter = nextResponse.next();
+            nextExpectedOffset += nextSize;
+            remaining -= nextSize;
+        }
```

Comment:
Can we assert `remaining == 0` here?

Reply:
We can, see 58ff48ce1f95df37fb27efc0587d3c6663650807.